tests/docker/test-helpers.go

```diff
@@ -0,0 +1,480 @@
+	config := &TestConfig{
+		K3sImage: k3sImage,
+	}
+
+	// Create temporary directory
+	tempDir, err := os.MkdirTemp("", "k3s-test-")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temp directory: %v", err)
+	}
+	config.TestDir = tempDir
+	// Setup cleanup on exit
+	// setupCleanup(config)
+
+	// Create required directories
+	if err := os.MkdirAll(filepath.Join(config.TestDir, "logs"), 0755); err != nil {
+		return nil, fmt.Errorf("failed to create logs directory: %v", err)
+	}
+
+	// Generate random secret
+	config.Secret = fmt.Sprintf("%012d", rand.Int63n(1000000000000))
+	return config, nil
+}
+
+// portUsed checks if a port is in use
+func portFree(port int) bool {
+	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		return false
+	}
+	listener.Close()
+	return true
+}
+
+// getPort finds an available port
+func getPort() int {
+	var port int
+	for i := 0; i < 100; i++ {
+		port = 10000 + rand.Intn(50000)
+		if portFree(port) {
+			return port
+		}
+	}
+	return -1
+}
+
+// ProvisionServers starts the required number of k3s servers
+// and updates the kubeconfig file with the first cp server details
+func (config *TestConfig) ProvisionServers(numOfServers int) error {
+	config.NumServers = numOfServers
+	for i := 0; i < config.NumServers; i++ {
+
+		// If a server i already exists, skip. This is useful for scenarios where
+		// the first server is started seperate from the rest of the servers
+		if config.Servers != nil && i < len(config.Servers) {
+			continue
+		}
+
+		testID := filepath.Base(config.TestDir)
+		name := fmt.Sprintf("k3s-server-%d-%s", i, strings.ToLower(testID))
+
+		port := getPort()
+		if port == -1 {
+			return fmt.Errorf("failed to find an available port")
+		}
+
+		serverImage := getEnvOrDefault("K3S_IMAGE_SERVER", config.K3sImage)
+
+		var joinOrStart string
+		if numOfServers > 0 {
+			if i == 0 {
+				joinOrStart = "--cluster-init"
+			} else {
+				if config.Servers[0].URL == "" {
+					return fmt.Errorf("first server URL is empty")
+				}
+				joinOrStart = fmt.Sprintf("--server %s", config.Servers[0].URL)
+			}
+		}
+
+		// Assemble all the Docker args
+		dRun := strings.Join([]string{"docker run -d",
+			"--name", name,
+			"--hostname", name,
+			"--privileged",
+			"-p", fmt.Sprintf("127.0.0.1:%d:6443", port),
+			"-p", "6443",
+			"-e", fmt.Sprintf("K3S_TOKEN=%s", config.Secret),
+			"-e", "K3S_DEBUG=true",
+			os.Getenv("SERVER_DOCKER_ARGS"),
+			os.Getenv(fmt.Sprintf("SERVER_%d_DOCKER_ARGS", i)),
+			os.Getenv("REGISTRY_CLUSTER_ARGS"),
+			serverImage,
+			"server", joinOrStart, os.Getenv("SERVER_ARGS"), os.Getenv(fmt.Sprintf("SERVER_%d_ARGS", i))}, " ")
+		if out, err := RunCommand(dRun); err != nil {
+			return fmt.Errorf("failed to run server container: %s: %v", out, err)
+		}
+
+		// Get the IP address of the container
+		ipOutput, err := RunCommand("docker inspect --format \"{{ .NetworkSettings.IPAddress }}\" " + name)
```

Comment:
do we wanna get super fancy and just directly call the docker APIs here using the golang client, instead of wrapping the docker CLI? I guess that might be a bit much.

Reply:
I looked into using something like that, but it just seemed like more effort than was worth it at the moment. I would leave this for future PRs. Converting the existing tests basically "as-is" was more my goal with this.